src/lightning/fabric/strategies/xla_fsdp.py

```diff
@@ -0,0 +1,404 @@
+        raise NotImplementedError(
+            f"The `{type(self).__name__}` does not support the joint setup of module and optimizer(s)."
+            " Please do it in this order: Create the model, call `setup_module`, create the optimizer,"
+            " call `setup_optimizer`."
+        )
+
+    def setup_module(self, module: Module) -> Module:
+        from torch_xla.distributed.fsdp import XlaFullyShardedDataParallel as XLAFSDP
+
+        if any(isinstance(mod, XLAFSDP) for mod in module.modules()):
+            if "auto_wrap_policy" in self._fsdp_kwargs:
+                rank_zero_warn(
+                    "A XLAFSDP `auto_wrap_policy` is set, but the model is already wrapped. The policy will be ignored."
+                )
+                del self._fsdp_kwargs["auto_wrap_policy"]
+        else:
+            if self._sync_module_states:
+                import torch_xla.core.xla_model as xm
+
+                xm.broadcast_master_param(module)
+
+            module = XLAFSDP(
+                module=module,
+                **self._fsdp_kwargs,
+            )
+
+        return module
+
+    def setup_optimizer(self, optimizer: Optimizer) -> Optimizer:
+        """Set up an optimizer for a model wrapped with XLAFSDP.
+
+        This setup method doesn't modify the optimizer or wrap the optimizer. The only thing it currently does is verify
+        that the optimizer was created after the model was wrapped with :meth:`setup_module` with a reference to the
+        flattened parameters.
+        """
+        if _TORCH_GREATER_EQUAL_2_0:
+            return optimizer
+
+        from torch_xla.distributed.fsdp.xla_flatten_params_wrapper import FlatParameter
+
+        num_groups = len(optimizer.param_groups)
+        if num_groups > 1:
+            raise ValueError(
+                "An optimizer used with an XLA FSDP model does not support multiple param groups."
+                f" Found {num_groups} parameter groups."
+            )
+
+        if any(isinstance(param, FlatParameter) for param in optimizer.param_groups[0]["params"]):
+            return optimizer
+
+        raise ValueError(
+            "The optimizer does not seem to reference any XLA FSDP parameters. HINT: Make sure to create the optimizer"
+            " after setting up the model."
+        )
+
+    def optimizer_step(
+        self,
+        optimizer: Optimizable,
+        **kwargs: Any,
+    ) -> Any:
+        """Overrides default tpu optimizer_step since FSDP should not call
+        `torch_xla.core.xla_model.optimizer_step`. Performs the actual optimizer step.
+
+        Args:
+            optimizer: the optimizer performing the step
+            **kwargs: Any extra arguments to ``optimizer.step``
+        """
+        loss = optimizer.step(**kwargs)
+        import torch_xla.core.xla_model as xm
+
+        xm.mark_step()
+        return loss
+
+    def module_to_device(self, module: Module) -> None:
+        module.to(self.root_device)
```

Comment:
You'll notice that regular FSDP makes this a no-op.

Wouldn't this be a problem here too?

Reply:
Hm... FSDPStrategy inherits this from the Strategy base class, which by default defines module_to_device as a no-op so I think this function isn't needed in the FSDP class: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/strategies/strategy.py#L160

The module_to_device in XLAFSDPStrategy also is not needed since it is the same in XLAStrategy. I think since we want to move our model to TPU to run calculations, it makes sense keep module_to_device as an op.

Reply:
Let me know if you agree here, for the time being will remove this function since it is the same in the parent class `XLAStrategy`

Reply:
Okay I looked into this further. The `XlaFullyShardedDataParallel` wrapper will move the model to the xla device if it has not been moved already. From this, it doesn't matter when or not we manually move the model.

However, In `setup_module()` we have the following:
```
if self._sync_module_states:
                if _XLA_GREATER_EQUAL_2_1:
                    from torch_xla.core.xla_model import broadcast_master_param
                else:
                    from torch_xla.experimental.pjrt import broadcast_master_param

                broadcast_master_param(module)

            module = XLAFSDP(
                module=module,
                **self._fsdp_kwargs,
            )
```

Now, `broadcast_master_param()` requires the module to already be on a `xla` device. I think it makes more sense to do `broadcast_master_param()` before wrapping around `XLAFSDP` (which is what we do now). If we leave the ordering as is, then we need to keep 

```
def module_to_device(self, module: Module) -> None:
        module.to(self.root_device)
```

Reply:
cc @carmocca

Reply:
SGTM